torch/_inductor/codegen/memory_planning.py

```diff
@@ -0,0 +1,798 @@
+        self.pool = pool
+        self.offset = offset
+        return self
+
+    def codegen_alloc_from_pool(self, wrapper):
+        node = self.node
+        shape = tuple(node.get_size())
+        stride = tuple(node.get_stride())
+        return "alloc_from_pool({})".format(
+            ", ".join(
+                map(
+                    str,
+                    [
+                        self.pool.name,
+                        pexpr(self.offset),  # bytes not numel
+                        node.get_dtype(),
+                        wrapper.codegen_shape_tuple(shape),
+                        wrapper.codegen_shape_tuple(stride),
+                    ],
+                )
+            )
+        )
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__}("
+            f"node={self.node.get_name()}, "
+            f"live_Range={self.live_range}, "
+            f"size_hint={self.size_hint}, "
+            f"symbolic_size={self.symbolic_size}, "
+            f"pool={self.pool.name}, "
+            f"offset={self.offset})"
+        )
+
+
+@dataclasses.dataclass
+class Empty(AllocationTreeNode):
+    """
+    Placeholder to represent empty space in the allocation pool.
+    Only exists to get the size_hint correct in parent nodes.
+    """
+
+    size_hint: int
+
+    def get_live_ranges(self):
+        return LiveRanges([])
+
+    def get_size_hint(self):
+        return self.size_hint
+
+    def get_symbolic_size(self):
+        return 0
+
+    def is_empty(self):
+        return True
+
+
+class ClearCacheOnAllocateMixin:
+    """
+    Helper to assist in caching get_live_ranges, get_size_hint, and
+    get_symbolic_size.
+    """
+
+    def allocate(self, block: "Allocation", is_last: bool):
+        is_allocated = self._allocate(block, is_last)
+        if is_allocated:
+            self.clear_cache()
+        return is_allocated
+
+    def clear_cache(self):
+        self.get_live_ranges.clear_cache(self)
+        self.get_size_hint.clear_cache(self)
+        self.get_symbolic_size.clear_cache(self)
+
+
+@dataclasses.dataclass
+class TemporalSplit(ClearCacheOnAllocateMixin, AllocationTreeNode):
+    """
+    Contains a list of allocations not overlapping in LiveRanges.
+
+    Invariant: no pair (a,b) in self.allocations will have:
+         a.get_live_ranges().overlaps(b.get_live_ranges())
+    """
+
+    allocations: List[AllocationTreeNode]
+
+    def _allocate(self, block: "Allocation", is_last: bool):
+        slot_size = self.get_size_hint()
+        block_size = block.get_size_hint()
+        if not is_last and block_size > slot_size:
+            return False  # doesn't fit
+
+        block_live = block.get_live_ranges()
+        overlapping = [
+            s for s in self.allocations if s.get_live_ranges().overlaps(block_live)
+        ]
+        if len(overlapping) > 1:
+            # TODO(jansel): we could try harder here by merging overlapping in space
```

Comment:
It seems that it's still possible to allocate the block if there is enough remaining space?

Reply:
FYI this is the memory planner in AIT: https://github.com/facebookincubator/AITemplate/blob/main/python/aitemplate/compiler/transform/memory_planning.py#L196, which tries to fill in gaps between different allocations.

Reply:
This is using roughly the same algorithm as the AIT one.  They both do greedy by size.

The difficulty here, and the reason for the tree, is dealing with dynamic shapes.  Which the original paper on this doesn't support.  If you have one tensor of size `s0` and another of size `s1` you don't nessarily know which one of those is bigger.  The tree allows us to compute symbolic boundaries (e.g. `align(max(s0, s1))`) that will turn into runtime dynamic shape expressions.

This specific line of code isn't actually hit on the models I tested, so doesn't matter.